vulkan: implement initial support for IQ2 and IQ3 quantizations

remyoudompheng opened the pull request:
This pull request implements basic support for IQ2 and IQ3 quantizations in the Vulkan backend, with tentative acceptable performance (there are probably possible improvements). Unfortunately I do not have access to coopmat2 hardware and there might be typos in the proposed implementation.

A commit modifies the `Q3_K` implementation to optimize performance, but it may be unwelcome in this PR.

The existing `init_iq4nl_shmem` function has been renamed to a more generic name in order to simplify `ifdef` logic.

Tests were performed on a Radeon 780M iGPU with Mesa 24.3.3 using the default compiler (ACO, not LLVM). It supports `KHR_coopmat`.

Performance results:
```
ggml_vulkan: 0 = AMD Radeon 780M (RADV GFX1103_R1) (radv) | uma: 1 | fp16: 1 | warp size: 64 | matrix cores: KHR_coopmat
| model                          |       size |     params | backend    | ngl |          test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------------: | -------------------: |
| qwen2 3B IQ2_M - 2.7 bpw       |   1.06 GiB |     3.09 B | Vulkan     |  99 |         pp512 |        479.81 ± 0.67 |
| qwen2 3B IQ2_M - 2.7 bpw       |   1.06 GiB |     3.09 B | Vulkan     |  99 |         tg128 |         26.03 ± 0.08 |
| qwen2 3B IQ3_XS - 3.3 bpw      |   1.29 GiB |     3.09 B | Vulkan     |  99 |         pp512 |        516.05 ± 0.81 |
| qwen2 3B IQ3_XS - 3.3 bpw      |   1.29 GiB |     3.09 B | Vulkan     |  99 |         tg128 |         39.28 ± 0.12 |
| qwen2 3B IQ3_S mix - 3.66 bpw  |   1.38 GiB |     3.09 B | Vulkan     |  99 |         pp512 |       509.56 ± 14.44 |
| qwen2 3B IQ3_S mix - 3.66 bpw  |   1.38 GiB |     3.09 B | Vulkan     |  99 |         tg128 |         35.83 ± 0.55 |
| qwen2 3B Q3_K - Small          |   1.35 GiB |     3.09 B | Vulkan     |  99 |         pp512 |        502.33 ± 0.74 |
| qwen2 3B Q3_K - Small          |   1.35 GiB |     3.09 B | Vulkan     |  99 |         tg128 |         43.28 ± 0.54 |
| qwen2 3B Q3_K - Medium         |   1.48 GiB |     3.09 B | Vulkan     |  99 |         pp512 |        504.33 ± 0.63 |
| qwen2 3B Q3_K - Medium         |   1.48 GiB |     3.09 B | Vulkan     |  99 |         tg128 |         40.19 ± 0.33 |
| qwen2 3B Q4_K - Medium         |   1.79 GiB |     3.09 B | Vulkan     |  99 |         pp512 |        509.78 ± 1.15 |
| qwen2 3B Q4_K - Medium         |   1.79 GiB |     3.09 B | Vulkan     |  99 |         tg128 |         34.43 ± 0.19 |
```

Performance numbers from `test-backend-ops`
```
  MUL_MAT(type_a=q2_K,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                   4260 runs -   268.20 us/run - 117.44 MFLOP/run - 437.89 GFLOPS
  MUL_MAT(type_a=q3_K,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                   2556 runs -   541.34 us/run - 117.44 MFLOP/run - 216.94 GFLOPS
  MUL_MAT(type_a=q4_K,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                   2556 runs -   431.30 us/run - 117.44 MFLOP/run - 272.29 GFLOPS
  MUL_MAT(type_a=iq2_xxs,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                3408 runs -   344.17 us/run - 117.44 MFLOP/run - 341.23 GFLOPS
  MUL_MAT(type_a=iq2_xs,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                 3408 runs -   365.16 us/run - 117.44 MFLOP/run - 321.62 GFLOPS
  MUL_MAT(type_a=iq2_s,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                  2556 runs -   490.09 us/run - 117.44 MFLOP/run - 239.63 GFLOPS
  MUL_MAT(type_a=iq3_xxs,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                3408 runs -   330.88 us/run - 117.44 MFLOP/run - 354.93 GFLOPS
  MUL_MAT(type_a=iq3_s,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                  2556 runs -   392.55 us/run - 117.44 MFLOP/run - 299.18 GFLOPS
  MUL_MAT(type_a=iq4_nl,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                 2556 runs -   427.37 us/run - 117.44 MFLOP/run - 274.80 GFLOPS
```

Before `Q3_K` change:
```
  MUL_MAT(type_a=q3_K,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                   2556 runs -   541.34 us/run - 117.44 MFLOP/run - 216.94 GFLOPS
  MUL_MAT(type_a=q3_K,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                   34 runs - 29597.18 us/run -  60.13 GFLOP/run -   2.03 TFLOPS
```
After `Q3_K` change
```
  MUL_MAT(type_a=q3_K,type_b=f32,m=4096,n=1,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                   3408 runs -   337.65 us/run - 117.44 MFLOP/run - 347.82 GFLOPS
  MUL_MAT(type_a=q3_K,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                   52 runs - 19365.63 us/run -  60.13 GFLOP/run -   3.10 TFLOPS
```

remyoudompheng committed:
vulkan: initial support for IQ3_S

remyoudompheng committed:
vulkan: initial support for IQ3_XXS

remyoudompheng committed:
vulkan: initial support for IQ2_XXS

remyoudompheng committed:
vulkan: initial support for IQ2_XS

remyoudompheng committed:
vulkan: optimize Q3_K by removing branches

remyoudompheng committed:
vulkan: implement dequantize variants for coopmat2

remyoudompheng committed:
vulkan: initial support for IQ2_S

remyoudompheng committed:
vulkan: vertically realign code

0cc4m commented:
Thank you, very cool! This will take a bit of time to review, I'll take a look this weekend. Can you fix the conflict? The mmq_wg_denoms fix has to be applied here too: #11343

jeffbolznv commented:
Exciting to see this. I've done a quick check with coopmat2 and there are a few failures:

```
  MUL_MAT_ID(type_a=iq2_s,type_b=f32,n_mats=4,n_used=2,b=0,m=512,n=32,k=256): [MUL_MAT_ID] NMSE = 0.083714476 > 0.000500000 FAIL
  MUL_MAT_ID(type_a=iq3_xxs,type_b=f32,n_mats=4,n_used=2,b=0,m=512,n=32,k=256): [MUL_MAT_ID] NMSE = 1.604342081 > 0.000500000 FAIL
  MUL_MAT_ID(type_a=iq3_s,type_b=f32,n_mats=4,n_used=2,b=0,m=512,n=32,k=256): [MUL_MAT_ID] NMSE = 0.533599553 > 0.000500000 FAIL
```

Not seeing failures with coopmat1 or no coopmat. I'll try to debug these later. I haven't looked at the code yet.

jeffbolznv commented:
I was surprised it was only the MUL_MAT_ID tests failing, but it was due to a gap in test coverage, which https://github.com/ggerganov/llama.cpp/pull/11375 will fix. MUL_MAT also fails for coopmat2 with the same types.

jeffbolznv committed:
port failing dequant callbacks from mul_mm

jeffbolznv commented:
I went ahead and did the straightforward unoptimized "port" of the failing dequant callbacks from mul_mm.comp - just divide the index by 2 (because mul_mm does pairs at a time) and replace data_a with the block reference. Code is at https://github.com/ggerganov/llama.cpp/commit/078ebe536bc3d64f14c3e215e5d83a61e32bb595. Feel free to pull this in however you want. IMO it would be OK to have these be unoptimized at first and you or I can optimize the later. I haven't done any perf testing yet.

remyoudompheng commented:
Thanks for the comments, I rebased the branch to include https://github.com/ggerganov/llama.cpp/pull/11343 and cherry-picked https://github.com/ggerganov/llama.cpp/commit/078ebe536bc3d64f14c3e215e5d83a61e32bb595

jeffbolznv commented on ggml/src/ggml-vulkan/vulkan-shaders/types.comp:
```diff
@@ -294,6 +294,748 @@ struct block_q6_K_packed16
+{
+    // copy the table into shared memory and sync
+    if (gl_LocalInvocationIndex.x < 32) {
+        for (uint i = gl_LocalInvocationIndex.x; i < 512; i += 32) {
```
Also, the increment could use gl_WorkGroupSize.x instead of hardcoding 32, but it probably won't affect performance much in practice.

remyoudompheng replied:
The loop looks better using workgroup size and less error prone, I made the fix

jeffbolznv commented on ggml/src/ggml-vulkan/vulkan-shaders/types.comp:
```diff
@@ -294,6 +294,748 @@ struct block_q6_K_packed16
+    uvec2(0x082b082b, 0x2b2b2b2b), uvec2(0x082b2b08, 0x2b2b2b2b), uvec2(0x2b082b08, 0x2b2b2b2b), uvec2(0x2b2b2b2b, 0x2b2b2b2b)
+};
+
+shared uvec2 iq2s_grid[1024];
```
We should probably account for these array sizes in ggml_vk_matmul_shmem_support. It's a bug that we didn't for iq4_nl, but these are significantly larger.

jeffbolznv commented on ggml/src/ggml-vulkan/vulkan-shaders/types.comp:
```diff
@@ -294,6 +294,748 @@ struct block_q6_K_packed16
+    0x3e1c1c1c, 0x3e1c3404, 0x3e24140c, 0x3e24240c, 0x3e2c0404, 0x3e2c0414, 0x3e2c1424, 0x3e341c04,
+};
+
+shared uint32_t iq3xxs_grid[512];
```
array sizes don't match

remyoudompheng replied:
Fixed

jeffbolznv reviewed (commented):
I didn't review that actual dequantization logic, but I reviewed the rest. And I still need to do some perf testing.

jeffbolznv commented:
I've been working on optimizing the cm2 dequant callbacks and am getting good speedups. I'm out of time to finish it tonight, but I'll share it in the morning.

jeffbolznv commented:
Here are the cm2 optimizations: https://github.com/jeffbolznv/llama.cpp/commit/9079f06e6b6bd3e9ff2b5f8957c7a1d6c84f665d

RTX 4070 (including Q4_K for reference):
```
before:
  MUL_MAT(type_a=q4_K,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                  616 runs -  1625.19 us/run -  60.13 GFLOP/run -  37.00 TFLOPS
  MUL_MAT(type_a=iq2_xxs,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):               294 runs -  3402.81 us/run -  60.13 GFLOP/run -  17.67 TFLOPS
  MUL_MAT(type_a=iq2_xs,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                308 runs -  3250.11 us/run -  60.13 GFLOP/run -  18.50 TFLOPS
  MUL_MAT(type_a=iq2_s,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                 324 runs -  3099.17 us/run -  60.13 GFLOP/run -  19.40 TFLOPS
  MUL_MAT(type_a=iq3_xxs,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):               378 runs -  2656.58 us/run -  60.13 GFLOP/run -  22.63 TFLOPS
  MUL_MAT(type_a=iq3_s,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                 438 runs -  2291.55 us/run -  60.13 GFLOP/run -  26.24 TFLOPS  
  
after:
  MUL_MAT(type_a=q4_K,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                  632 runs -  1583.83 us/run -  60.13 GFLOP/run -  37.96 TFLOPS
  MUL_MAT(type_a=iq2_xxs,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):               596 runs -  1682.00 us/run -  60.13 GFLOP/run -  35.75 TFLOPS
  MUL_MAT(type_a=iq2_xs,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                608 runs -  1647.64 us/run -  60.13 GFLOP/run -  36.49 TFLOPS
  MUL_MAT(type_a=iq2_s,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                 490 runs -  2044.33 us/run -  60.13 GFLOP/run -  29.41 TFLOPS
  MUL_MAT(type_a=iq3_xxs,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):               682 runs -  1467.80 us/run -  60.13 GFLOP/run -  40.97 TFLOPS
  MUL_MAT(type_a=iq3_s,type_b=f32,m=4096,n=512,k=14336,bs=[1,1],nr=[1,1],per=[0,1,2,3]):                 672 runs -  1490.99 us/run -  60.13 GFLOP/run -  40.33 TFLOPS
```

remyoudompheng committed:
Fix array length mismatches

remyoudompheng commented:
Branch updated:
* fix array lengths issues
* rebase to resolve conflict with #11375 
* no cherry-pick of https://github.com/jeffbolznv/llama.cpp/commit/9079f06e6b6bd3e9ff2b5f8957c7a1d6c84f665d
* no change to shmem checks (they seem quite complex)

remyoudompheng commented:
For some reason commit https://github.com/ggerganov/llama.cpp/pull/11360/commits/6ed3047e4121ca9922d6addab6cb32a4b48bd2c0 causes a huge performance regression on my device (Radeon 780M using Mesa ACO compiler or LLVM compiler) as if all benefits from the shared array were lost.

I restored the hardcoded 32 value in https://github.com/ggerganov/llama.cpp/pull/11360/commits/3f7aa9dcb13c9903d77af1b42992e03f14193963

0cc4m reviewed (approved):
I don't see any issues on AMD, Intel or Nvidia in my tests. Performance isn't that good yet, but that can be improved separately.

netrunnereve commented on ggml/src/ggml-vulkan/vulkan-shaders/types.comp:
```diff
@@ -294,6 +294,748 @@ struct block_q6_K_packed16
+void init_iq_shmem()
+{
+    // copy the table into shared memory and sync
+    if (gl_LocalInvocationIndex.x < 32) {
```
For the shaders that have a variable workgroup size this will cause issues with devices that have a subgroup size smaller than 32. You'll need to do something like what I did in https://github.com/ggerganov/llama.cpp/pull/10809 to fix this.

I have a feeling that this might be what's causing the llvmpipe ci failures as well.

jeffbolznv replied:
It's too bad the version using gl_WorkgroupSize didn't work, because it wouldn't have had that issue. I wonder if writing the loop like this would avoid the perf regression:

```
    [[unroll]] for (uint i = 0; i < iq2xxs_grid.length(); i += gl_WorkGroupSize.x) {
        iq2xxs_grid[i + gl_LocalInvocationIndex.x] = iq2xxs_grid_const[i + gl_LocalInvocationIndex.x];
    }
```

This should fully unroll and will work without additional branches as long as the workgroupsize evenly divides the array size.

netrunnereve commented:
Well you beat me to this :wink:

As @0cc4m mentioned I think that speed is not really a priority here and the important part is to have a functional implementation we can improve on in the future. This will probably need dedicated mat vec shaders like the K quants to get full performance.

Please see my review but aside from that your code looks fine (I didn't verify the actual dequantization algorithm though) and runs fine on my AMD GCN cards. If you want to be sure that your implementation is correct it's worth running a perplexity check against a different backend to see if the numbers match up.

remyoudompheng commented:
The bug regarding `gl_WorkgroupSize` happens at the `glslc` step, because even the SPIR-V looks wrong (even with `[[unroll]]`) the loop is compiled with a stride of 1, making each shader invocation very costly.

```
$ glslc  -DDATA_A_IQ2_XS=1 -DB_TYPE=float -DB_TYPE_VEC2=vec2 -DB_TYPE_VEC4=vec4 -DD_TYPE=float mul_mat_vec.comp -DFLOAT_TYPE=float -o mul_mat_vec_iq2xs.spv
$ spirv-dis mul_mat_vec_iq2xs.spv
...
               ; Function init_iq_shmem_
%init_iq_shmem_ = OpFunction %void None %7

         %11 = OpLabel
        %i_0 =   OpVariable %_ptr_Function_uint Function
  %indexable =   OpVariable %_ptr_Function__arr_v2uint_uint_512 Function
                 OpLine %4 553 0
                 OpStore %i_0 %uint_0
                 OpBranch %65

         %65 = OpLabel
                 OpLine %4 553 0
                 OpLoopMerge %67 %68 None
                 OpBranch %69

         %69 =     OpLabel
                     OpLine %4 553 0
         %70 =       OpLoad %uint %i_0
         %72 =       OpULessThan %bool %70 %uint_512
                     OpBranchConditional %72 %66 %67

         %66 =         OpLabel
                         OpLine %4 554 0
         %77 =           OpLoad %uint %i_0
         %80 =           OpLoad %uint %gl_LocalInvocationIndex
         %81 =           OpIAdd %uint %77 %80
        %676 =           OpLoad %uint %i_0
        %677 =           OpLoad %uint %gl_LocalInvocationIndex
        %678 =           OpIAdd %uint %676 %677
                         OpStore %indexable %675
        %682 =           OpAccessChain %_ptr_Function_v2uint %indexable %678
        %683 =           OpLoad %v2uint %682
        %685 =           OpAccessChain %_ptr_Workgroup_v2uint %iq2xs_grid %81
                         OpStore %685 %683
                         OpBranch %68

         %68 =   OpLabel
                   OpLine %4 553 0
        %687 =     OpLoad %uint %i_0
        %688 =     OpIAdd %uint %687 %uint_1
                   OpStore %i_0 %688
                   OpBranch %65
```
(notice the `%uint_1` on instruction 688, it is `%uint_32` when 32 is hardcoded)

It seems to be this issue: https://github.com/KhronosGroup/glslang/issues/2479

The workaround is to include `types.comp` *after* the workgroup size is declared in the shader. If this is fine to you I can update the PR to use this method (meaning minor changes to all shaders to move the include, or at least the ones calling `init_iq_shmem`)

remyoudompheng committed:
vulkan: avoid using workgroup size before it is referenced

remyoudompheng commented:
PR updated:
* removed https://github.com/ggerganov/llama.cpp/pull/11360/commits/3f7aa9dcb13c9903d77af1b42992e03f14193963
* added https://github.com/ggerganov/llama.cpp/pull/11360/commits/edd2546dd3caae072296c6a68bc475781589ea7f

llvmpipe is now happy

jeffbolznv commented:
> The workaround is to include types.comp after the workgroup size is declared in the shader. 

I'm worried that somebody will accidentally break this in the future and it'll be very confusing. Another option might be to pass the workgroup size as a function parameter.

0cc4m commented:
I think a function parameter is a good idea, yeah.

sorasoras commented:
looking forward to IQ4XS support as well.

remyoudompheng commented:
Indeed, it looks better with a function parameter the branch is updated.

@sorasoras feel free to test branch https://github.com/remyoudompheng/llama.cpp/commit/f7a32d00dc3955880c03772e6033c933909811db

sorasoras commented:
> Indeed, it looks better with a function parameter the branch is updated.
> 
> @sorasoras feel free to test branch [remyoudompheng@e955cbed](https://github.com/remyoudompheng/llama.cpp/commit/e955cbed)

```
 .\llama-bench.exe -m W:\model\sakura-14b-qwen2beta-v0.9-IQ4_XS.gguf -ngl 99 -sm none
ggml_vulkan: Found 1 Vulkan devices:
ggml_vulkan: 0 = AMD Radeon RX 7900 XTX (AMD proprietary driver) | uma: 0 | fp16: 1 | warp size: 64 | matrix cores: KHR_coopmat
| model                          |       size |     params | backend    | ngl |    sm |          test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ----: | ------------: | -------------------: |
| qwen2 13B IQ4_XS - 4.25 bpw    |   7.37 GiB |    14.17 B | Vulkan     |  99 |  none |         pp512 |       1418.05 ± 2.73 |
| qwen2 13B IQ4_XS - 4.25 bpw    |   7.37 GiB |    14.17 B | Vulkan     |  99 |  none |         tg128 |         50.25 ± 0.29 |

build: e955cbed (4570)
```
it looks ok

remyoudompheng commented:
The test failure seems unrelated to patch correctness (llvmpipe is too slow for test-backend-ops to complete following the increase in test cases)

jeffbolznv commented:
Is it possible to increase the timeout? Or what do we need to do to unblock this change.

netrunnereve commented:
> Indeed, it looks better with a function parameter the branch is updated.

We should probably make the change for iq4_nl as well so things are consistent.

> Is it possible to increase the timeout?

Yeah you can change it here, and if it's still faster than the CUDA and HIP builds I don't think anyone will mind.

https://github.com/ggerganov/llama.cpp/blob/a4417ddda98fd0558fb4d802253e68a933704b59/.github/workflows/build.yml#L349

remyoudompheng committed:
tests: increase timeout for Vulkan llvmpipe backend

remyoudompheng commented:
Branch updated
* refactor `init_iq_shmem` for iq4nl (https://github.com/ggerganov/llama.cpp/pull/11360/commits/0886297a91347c0dc861ede41c4385f39003a081 replaced by https://github.com/ggerganov/llama.cpp/pull/11360/commits/aa17d321b36e2d8ab912c0b05e05088793221567)
* increase test timeout

netrunnereve commented:
So I tried out the new Github ARM CI machine for fun and that one actually finishes the llvmpipe test stage for master in only 7 minutes. The problem is that it has a subgroup size of 4 with a 128 bit wide SIMD and that's causing new test failures.

jeffbolznv commented:
> The problem is that it has a subgroup size of 4 with a 128 bit wide SIMD and that's causing new test failures.

Can you clarify what you mean by 128b wide SIMD, and whether these new test failures are specific to this PR?

IMO this change is complete other than accounting for the additional shared memory usage. I think it would be OK to merge even with that bug and then I or @remyoudompheng could fix it after.

0cc4m commented:
> > The problem is that it has a subgroup size of 4 with a 128 bit wide SIMD and that's causing new test failures.
> 
> Can you clarify what you mean by 128b wide SIMD, and whether these new test failures are specific to this PR?

I think that's about llvmpipe using SIMD width as subgroup size. For AVX2 it was 8, on ARM it is apparently 4. But that is unrelated to this PR.

> IMO this change is complete other than accounting for the additional shared memory usage. I think it would be OK to merge even with that bug and then I or @remyoudompheng could fix it after.

Yeah, I think it's okay to merge. We can clean up the shared memory checks in another PR. I think the checks got messed up in some refactor anyways and do some redundant work, your original version of the check and my later implementation currently coexist.

netrunnereve commented:
> I think that's about llvmpipe using SIMD width as subgroup size. For AVX2 it was 8, on ARM it is apparently 4. But that is unrelated to this PR.

Yep that's what I meant. I think we're fine to merge.

0cc4m commented:
Thank you for your contribution @remyoudompheng !

jeffbolznv commented:
FYI I've started implementing the shared memory accounting logic.

remyoudompheng commented:
I think some typos have slipped in the `dequant_iq*` shaders but it is very unclear how to spot them with `test-backend-ops` (see this diff for a few examples : 1b0739c44e993760ce9f2a15bb5a1df218480d00)
